Generate Timetable from TaskList and TaskFilter

daniellimws opened the pull request:
#### Summary of changes made to non-Timetable classes:
* `ImportanceFilter`: Create this class to filter tasks based on importance
* `TaskList`: Change `getFilteredTasks` to use `TaskFilter`
* `Task`: Change` importance` field type to `Importance` enum instead of `String`. Move `Task` class into `athena.task` package.
* `Parser` and `Command` classes: Update methods to use `Importance` enum
* `ListCommand`: Generate a timetable to be printed using `ui.printTimetable`

(Sorry all these were side effects due to using an `ImportanceFilter` class for filtering tasks)

* `Ui`: Uncomment `printTimetable`

#### Changes made to classes in `Timetable` package:
* Implement `toString()` methods for `Timetable` and `TimetableDay`
* `Timetable` can generate a list of `TimetableDay` objects from a `TaskList` and an optional `TaskFilter`.

Because of the way dates are parsed in `Timetable`, now we only can accept dates/recurrences (once off recurrence) in the `dd-MM-yyyy` format (Monday, Tuesday, ... not supported for now). The `LocalDate` class expects a year to be provided so I think we need to help the user add in the year in `Parser`. After this PR, I'll work on accepting days of the week (e.g. Monday, Tuesday, ...) for the recurrence instead of only allowing dates.

#### Tests
Also add tests to make sure `Timetable` correctly generates a list of `TimetableDay` objects from a `TaskList`, as well as from a `TaskList` filtered with an `ImportanceFilter` set to `HIGH` importance.

daniellimws committed:
Fix build and checkstyle errors

daniellimws committed:
Generate Timetable from TaskList and TaskFilter

Changes made to:
TaskList: Change getFilteredTasks to use TaskFilter
Task: Change importance to Importance enum
Parser and Command classes: Update methods to use Importance enum
Ui: Uncomment printTimetable

daniellimws committed:
Fix checkstyletest

daniellimws committed:
Fix runtime errors

To convert importance string to enum, the string must be in uppercase
As for now, date must follow dd-MM-yyyy format

daniellimws committed:
Remove redundant imports

amanda-chua reviewed (approved):
looks good

daniellimws committed:
Merge branch 'master' into timetable-filter

wish2023 commented on src/main/java/athena/Parser.java:
```diff
@@ -123,8 +124,9 @@ public static Command parseEditCommand(String taskInfo, int namePos, int timePos
-        Command command = new ListCommand(importance);
+        String importance = getParameterDesc(taskInfo, IMPORTANCE_DELIMITER, importancePos,
+                Importance.MEDIUM.toString());
+        Command command = new ListCommand(Importance.valueOf(importance.toUpperCase()));
```
Good that the importance is being converted to upper case here!

daniellimws replied:
Yea I tried without and the program might crash if the user doesn't give it in upper case.

wish2023 commented on src/main/java/athena/Athena.java:
```diff
@@ -16,13 +16,14 @@ public static void main(String[] args) {
             command.execute(taskList, athena.ui);
         }
-        command = parser.parse("list");
+        command = parser.parse("list i/HIGH");
```
Remember that the user will type "high" instead of "HIGH"

wish2023 commented on src/main/java/athena/TaskList.java:
```diff
@@ -75,14 +76,14 @@ public Task markTaskAsDone(int taskNumber) {
      */
     public void addTask(String name, String startTime, String duration,
-                        String deadline, String recurrence, String importance, String notes, int index) {
+                        String deadline, String recurrence, Importance importance, String notes, int index) {
 
         Task task = createTask(name, startTime, duration, deadline, recurrence, importance, notes, index);
         tasks.add(task);
     }
 
     public void addTask(String name, String startTime, String duration,
-                        String deadline, String recurrence, String importance, String notes) {
+                        String deadline, String recurrence, Importance importance, String notes) {
         this.maxIndex += 1;
         int index = maxIndex;
 
```
I think we can remove the first addTask() . The method right below this seems to have the same functionality in addition to assigning a unique index value

daniellimws replied:
I agree!

daniellimws replied:
Oh wait, one allows the programmer to choose the index while the other will automatically assign an index. But yea there's repeated code so I'll modify the second `addTask` to call the first `addTask`.

wish2023 commented on src/main/java/athena/commands/ListCommand.java:
```diff
@@ -22,7 +26,8 @@ public ListCommand(String importance) {
     public void execute(TaskList taskList, Ui ui) {
-        TaskList tasksFiltered = taskList.getFilteredList(taskImportance);
-        ui.printList(tasksFiltered.getTasks());
+        ImportanceFilter importanceFilter = new ImportanceFilter(taskImportance);
+        Timetable timetable = new Timetable(taskList, importanceFilter);
+        ui.printTimetable(timetable);
```
Noted that the user will now see the list from the timetable class only

wish2023 commented on src/main/java/athena/task/taskfilter/ImportanceFilter.java:
```diff
@@ -0,0 +1,24 @@
+import athena.Importance;
+import athena.task.Task;
+
+public class ImportanceFilter extends TaskFilter {
```
Looks great!

wish2023 reviewed (approved):
Looks great! It may be better for the task class to have the deadline and recurrence to make it easier for the timetable class. Let's talk about this during the meeting. 

It may be good if one of us make a recurrence class too since recurrence can be a day of the week, or a string.

daniellimws commented:
Thanks. I agree it would be better to have a `Recurrence` class so that we can support different types of values.

daniellimws committed:
Merge branch 'master' into timetable-filter

daniellimws committed:
TaskList: Remove repeated code in addTask

daniellimws committed:
Fix some errors resulted from previous merge

Also, modify save.csv because for now, recurrence date must contain year (dd-MM-yyyy format)
